server/src/main/java/com/cloud/ha/HighAvailabilityManagerImpl.java

```diff
@@ -682,29 +687,44 @@ public void cancelDestroy(VMInstanceVO vm, Long hostId) {
 
     protected Long destroyVM(final HaWorkVO work) {
         final VirtualMachine vm = _itMgr.findById(work.getInstanceId());
-        s_logger.info("Destroying " + vm.toString());
+        if (vm == null) {
+            s_logger.info("No longer can find VM " + work.getInstanceId() + ". Throwing away " + work);
+            work.setStep(Step.Done);
+            return null;
+        }
+        boolean expunge = VirtualMachine.Type.SecondaryStorageVm.equals(vm.getType())
+                || VirtualMachine.Type.ConsoleProxy.equals(vm.getType());
+        if (!expunge && VirtualMachine.State.Destroyed.equals(work.getPreviousState())) {
+            s_logger.info("VM " + vm.getUuid() + " already in " + vm.getState() + " state. Throwing away " + work);
+            work.setStep(Step.Done);
+            return null;
+        }
         try {
-            if (vm.getState() != State.Destroyed) {
-                s_logger.info("VM is no longer in Destroyed state " + vm.toString());
-                return null;
+            if (VirtualMachine.State.Running.equals(work.getPreviousState())) {
+                _itMgr.advanceStop(vm.getUuid(), true);
             }
-
-            if (vm.getHostId() != null) {
-                _itMgr.destroy(vm.getUuid(), false);
-                s_logger.info("Successfully destroy " + vm);
-                return null;
-            } else {
-                if (s_logger.isDebugEnabled()) {
-                    s_logger.debug(vm + " has already been stopped");
+            s_logger.info("Destroying " + vm.toString());
+            if (!VirtualMachine.State.Expunging.equals(work.getPreviousState())) {
+                s_logger.info("Destroying " + vm.getUuid());
+                if (VirtualMachine.Type.ConsoleProxy.equals(vm.getType())) {
+                    consoleProxyManager.destroyProxy(vm.getId());
+                } else if (VirtualMachine.Type.SecondaryStorageVm.equals(vm.getType())) {
+                    secondaryStorageVmManager.destroySecStorageVm(vm.getId());
+                } else {
```

Comment:
the total method has grown from 23 to 40 lines, can you factor the new bits out please.